citest/gcp_testing/gcp_api.py

```diff
@@ -52,26 +55,155 @@ def build_authenticated_service(name, version, scopes, credentials_path):
   return discovery.build(name, version, http=http)
 
 
-class GoogleAgentObservationFailureVerifier(jc.ObservationFailureVerifier):
+class HttpErrorPredicateResult(jp.PredicateResult):
+  """Result from an HttpErrorPredicate."""
+
+  @property
+  def value(self):
+    """The value the predicate was applied to."""
+    return self.__value
+
+  def __init__(self, valid, value, comment=None):
+    super(HttpErrorPredicateResult, self).__init__(valid, comment=comment)
+    self.__value = value
+
+  def __eq__(self, result):
+    return (super(HttpErrorPredicateResult, self).__eq__(result)
+            and self.__value == result.value)
+
+  def export_to_json_snapshot(self, snapshot, entity):
+    """Implements JsonSnapshotable interface."""
+    builder = snapshot.edge_builder
+    builder.make_output('Error Value', self.__value)
+    super(HttpErrorPredicateResult, self).export_to_json_snapshot(snapshot,
+                                                                  entity)
+
+class HttpErrorPredicate(jp.ValuePredicate):
+  """Determine if an object matches an expected HttpError."""
+
+  @property
+  def http_code(self):
+    """The expected HTTP code or None."""
+    return self.__http_code
+
+  @property
+  def contains_regex(self):
+    """Describes the expected response content or None."""
+    return self.__contains_regex
+
+  def __init__(self, http_code=None, contains_regex=None):
+    """Constructor.
+
+    Args:
+      http_code: [int] The expected http_code or None.
+      error_regex: [string] A regex for the expected contentor None.
+    """
+    if http_code is None and contains_regex is None:
+      raise ValueError('Needs at least an http_code or contains_regex.')
+
+    self.__http_code = http_code
+    self.__contains_regex = contains_regex
+
+  def __call__(self, value):
+    if not isinstance(value, HttpError):
+      return jp.PredicateResult(
+          valid=False,
+          comment='{0} != HttpType'.format(value.__class__.__name__))
+
+    message = ['HttpError']
+    sep = 'with'  # Separator between confirmation phrases.
+    if self.__http_code:
```

Comment:
`if self.__http_code is not None:` since it's explicitly set to `None` in the case where it doesn't exist.